### ISSUE TITLE
`tf.keras.losses.cosine_similarity` generates unreasonably large gradients for zero-masked & broadcasted tensor

### ISSUE THREAD

**Issue by charlielam0615** [tensorflow/tensorflow#56305]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

binary

### Tensorflow Version

v2.9.0-18-gd8ce9f9c301 2.9.1

### Custom Code

Yes

### OS Platform and Distribution

Linux Ubuntu 21.04

### Current Behaviour?

```shell
For a zero-masked, broadcasted tensor, autograd generates very large gradients for zero-masked parts.
```

### Standalone code to reproduce the issue

```shell
import tensorflow as tf

def test():
    a = tf.random.normal([3, 5])
    with tf.GradientTape(persistent=True) as tape:
        tape.watch(a)
        a = a * tf.constant([[1],[0],[0]], dtype=tf.float32)
        cos_score = tf.keras.losses.cosine_similarity(a[:,None], a[None,:])
        loss = tf.reduce_sum(cos_score)

    grad = tape.gradient(loss, a)
    print(grad)

test()
```

### Relevant log output

```shell
A sample output from the above code

tf.Tensor(
[[-5.9604645e-08  5.9604645e-08 -4.4703484e-08 -2.9802322e-08
   8.9406967e-08]
 [ 9.4966488e+05 -9.1792988e+05  6.5334194e+05  3.2174541e+05
  -1.3134559e+06]
 [ 9.4966488e+05 -9.1792988e+05  6.5334194e+05  3.2174541e+05
  -1.3134559e+06]], shape=(3, 5), dtype=float32)
```
</details>

**Comment by sushreebarsa** [comment]
@gadagashwini I was able to replicate the issue on colab, please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/c0183168fcec2f36957c6a72f9f755a7/56305.ipynb) for reference.
Thank you!

**Comment by cantonios** [comment]
Hmm... seems to be because you are overwriting `a`.  Creating an intermediate `b` seems to work around the issue:
```
def test():
    a = tf.random.normal([3, 5])
    with tf.GradientTape(persistent=True) as tape:
        tape.watch(a)
        b = a * tf.constant([[1],[0],[0]], dtype=tf.float32)
        cos_score = tf.keras.losses.cosine_similarity(b[:,None], b[None,:])
        loss = tf.reduce_sum(cos_score)

    grad = tape.gradient(loss, a)
    print(grad)
```

**Comment by tilakrayal** [comment]
@charlielam0615,
I tried to execute the mentioned ode with the alternative approach as the developer mentioned, and got the output that autograd generates small gradients for zero-masked parts. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/7273ef056b60d83b860c2da83b135416/untitled.ipynb). Thank you!
